src/api/statements/JSONLDObject/index.js

```diff
@@ -40,34 +40,39 @@ const merge = (field, fields, currentOutput, data) => {
     return currentOutput;
 };
 
-export default function JSONLDObject(data, feed) {
+export default async function JSONLDObject(data, feed) {
     if (this.isLast()) {
         feed.close();
         return;
     }
-    const fields = this.getParam('fields', {});
+    const fields = this.getParam('fields', []);
     const collectionClass = this.getParam('collectionClass', '');
     const characteristics = this.getParam('characteristics', {});
     const exportDataset = this.getParam('exportDataset', false) === 'true';
-    const output = fields.filter(f => f.cover === 'collection').reduce(
-        (currentOutput, field) => {
+    const output = await fields.filter(f => f.cover === 'collection').reduce(
+        async (previousPromise, field) => {
+            const currentOutput = await previousPromise;
             if (collectionClass) currentOutput['@type'] = collectionClass;
-            return merge(field, fields, currentOutput, data);
+            return await merge(field, fields, currentOutput, data);
         },
-        {
+        Promise.resolve({
             '@id': getUri(data.uri),
-        }
+        })
     );
 
     if (this.isFirst() && exportDataset) {
-        output.dataset = fields
+        // https://gyandeeps.com/array-reduce-async-await/ TODO: write a test
+        output.dataset = await fields
```

Comment:
use common/lib/composeAsync, to compose async function
First you will also have to refactor merge to have the signature : 
```js
const merge = currentOutput => async (field, fields, characteristic) => ...
```

this way you can create an array of async operation to compose.
```js
const mergeOperations = fields.filter(...)
    .map(field => merge(field, fields, characteristics[0));
const output = await composeAsync(() => ({}), ...mergeOperations);
```
`() => ({})` is to get the first value.

Reply:
I must admit I didn't get the way it works.

I changed the prototype of `merge` to:

```js
const merge = currentOutput => async (field, fields, data) => {
```
(I did not change `data` to `characteristics` yet, one thing at a time)

And I tried:
```js
        const mergeOperations = fields
            .filter(f => f.cover === 'dataset')
            .map(field => merge(field, fields, characteristics[0]));
        output.dataset = await composeAsync(() => ({}), ...mergeOperations);
```
without success: 6 tests failed, including one in `src/api/exporters/exportJsonld.spec.js`. My tests failed the same way they did before I change the reduce part to take the asynchronous part into account.

I must admit I do not get the way `composeAsync` destructures its first argument, `funcs`. I don't understand what `...mergeOperations` should produce either.

Reply:
If I understand well, `mergeOperations` contains an array of Promise.
Oh, so, the `...mergeOperations` in the call of `composeAsync` only inline the promises as parameters of `composeAsync`. OK.

Reminder:
```js
const composeAsync = (...funcs) => async arg =>
    funcs.reduce((prevRes, fn) => prevRes.then(fn), Promise.resolve(arg));
```

Here, `funcs` will be an array of the first argument `() => ({})`, followed by the promises that were in the array `mergeOperations`.
Thus, `composeAsync` returns a function taking one argument `arg` and returning the result of a reduce, taking as a first accumulator value the resolved promise from `arg`.
What I still don't get is *where does `arg` come from?*

I'm not sure `composeAsync` as it is is adapted to what I want to do (consider that the second parameter of my reduce is `field`, whereas in `composeAsync`, it seems to be a function `fn`).

Reply:
I am now sure that `composeAsync` is not what I need.
I need a kind of async reduce (and my function is not curryfied, which prevents it from being a parameter to an helper).